### PR TITLE
Enforce status check for corruption_test

### PR DESCRIPTION
**Pull request by riversand963**
Enforce status check for corruption_test.

Test plan:
```
ASSERT_STATUS_CHECKED=1 make corruption_test
./corruption_test
```

### CODE REVIEW

File: db/db_impl/db_impl_compaction_flush.cc [hunk 1]
```diff
@@ -2982,9 +2983,10 @@ Status DBImpl::BackgroundCompaction(bool* made_progress,
       auto err_reason = versions_->io_status().ok()
                             ? BackgroundErrorReason::kCompaction
                             : BackgroundErrorReason::kManifestWrite;
-      error_handler_.SetBGError(io_s, err_reason);
+      error_handler_.SetBGError(io_s, err_reason).PermitUncheckedError();
```
Comment (mrambacher) [inline db/db_impl/db_impl_compaction_flush.cc:2986]: If (almost) everywhere that calls SetBGError just ignores the error, maybe the method should be changed to return a void?  Or there should be a second signature that returns a void?

File: db/table_cache.cc [hunk 2]
```diff
@@ -171,6 +171,7 @@ Status TableCache::FindTable(const ReadOptions& ro,
 
   if (*handle == nullptr) {
     if (no_io) {  // Don't do IO and return a not-found status
+      s.PermitUncheckedError();
```
Comment (mrambacher) [inline db/table_cache.cc:174]: Can you fix the comment above about "Not Found',  It returns Incomplete"

File: db/corruption_test.cc [hunk 3]
```diff
@@ -151,6 +153,7 @@ class CorruptionTest : public testing::Test {
         correct++;
       }
     }
+    iter->status().PermitUncheckedError();
```
[the remark this answers is not in the document]
Comment (riversand963) [reply db/corruption_test.cc:156]: I thought `PermitUncheckedError()` is for the purpose of temporarily letting some of the tests pass. If I put it in destructor of `CompactionIterator`, then it seems to mean callers of `CompactionIterator` does not have to check its `status_`. By letting caller determine whether and where to check `status_`, we can better enforce the checking. Let me know if I have misunderstood the intention.

File: db/db_impl/db_impl_compaction_flush.cc (hunk 1 above)
Comment (riversand963) [reply db/db_impl/db_impl_compaction_flush.cc:2986]: Good point. cc @anand1976

File: db/corruption_test.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (mrambacher) [reply db/corruption_test.cc:156]: I think there is generally an issue with Status as member variables.  If I create a CompactionIterator and never do anything with it, I will get an error about the status not being checked.  This seems wrong to me as I never did anything with the iterator. 

In any case, my guess is someone else is going to hit another case where the CompactionIterator status is not checked as well...

File: db/table_cache.cc (hunk 2 above)
Comment (riversand963) [reply db/table_cache.cc]: Done.

File: db/corruption_test.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (riversand963) [reply db/corruption_test.cc:156]: In the case you mentioned, I think it is easy to just call `PermitUncheckedError()` for the member `status`. This seems a little more, but not the worst thing. It will be good if this can be simplified.

[the remark this answers is not in the document]
Comment (riversand963) [reply db/corruption_test.cc:156]: Things can be more tedious if the class has multiple status members...

File: db/version_set.cc [hunk 4]
```diff
@@ -4137,7 +4137,8 @@ Status VersionSet::ProcessManifestWrites(
                      "\n",
                      manifest_file_number_, pending_manifest_file_number_);
       env_->DeleteFile(
-          DescriptorFileName(dbname_, pending_manifest_file_number_));
+              DescriptorFileName(dbname_, pending_manifest_file_number_))
+          .PermitUncheckedError();
```
Comment (jay-zhuang) [inline db/version_set.cc:4141]: should it log a error if a file failed to delete?

File: table/table_factory.cc [hunk 5]
```diff
@@ -37,7 +37,8 @@ Status TableFactory::CreateFromString(const ConfigOptions& config_options_in,
     factory->reset(new CuckooTableFactory());
 #endif  // ROCKSDB_LITE
   } else {
-    return Status::NotSupported("Could not load table factory: ", name);
+    status = Status::NotSupported("Could not load table factory: ", name);
+    return status;
```
Comment (jay-zhuang) [inline table/table_factory.cc:41]: Why this change is needed?

**Review by jay-zhuang — APPROVED** [review]
LGTM. Just a few minor questions.

File: db/write_thread.h [hunk 6]
```diff
@@ -83,6 +83,8 @@ class WriteThread {
     std::atomic<size_t> running;
     size_t size = 0;
 
+    ~WriteGroup() { status.PermitUncheckedError(); }
```
Comment (siying) [inline db/write_thread.h:86]: Can you add a comment about why this can be permitted?

File: db/db_impl/db_impl_compaction_flush.cc (hunk 1 above)
Comment (zhichao-cao) [reply db/db_impl/db_impl_compaction_flush.cc]: @anand1976 Mark suggested to have 2 versions of `SetBGError`. One returns the status after we deal with the BG error, and in this case, the caller will make some actions based on the returned error. Another one is a `void` method, we just set the BG error

File: db/version_set.cc (hunk 4 above)
Comment (riversand963) [reply db/version_set.cc:4141]: Good catch. Will update.

File: table/table_factory.cc (hunk 5 above)
Comment (riversand963) [reply table/table_factory.cc:41]: If we reach here and proceed, the `status` obj created in line 17 will be unchecked.

File: db/compaction/compaction_job.cc [hunk 7]
```diff
@@ -1062,6 +1063,8 @@ void CompactionJob::ProcessKeyValueCompaction(SubcompactionState* sub_compact) {
   }
   if (status.ok()) {
     status = c_iter->status();
+  } else {
+    c_iter->status().PermitUncheckedError();
```
[the remark this answers is not in the document]
Comment (riversand963) [reply db/compaction/compaction_job.cc:1067]: Will remove the change here.

File: db/write_thread.h (hunk 6 above)
Comment (riversand963) [reply db/write_thread.h:86]: Actually, I think it's better not to call `PermitUncheckedError()` in destructor. Let me update the code.

File: db/db_impl/db_impl_write.cc [hunk 8]
```diff
@@ -999,8 +999,9 @@ WriteBatch* DBImpl::MergeBatch(const WriteThread::WriteGroup& write_group,
     merged_batch = tmp_batch;
     for (auto writer : write_group) {
       if (!writer->CallbackFailed()) {
-        WriteBatchInternal::Append(merged_batch, writer->batch,
-                                   /*WAL_only*/ true);
+        Status s = WriteBatchInternal::Append(merged_batch, writer->batch,
+                                              /*WAL_only*/ true);
+        assert(s.ok());
```
Comment (siying) [inline db/db_impl/db_impl_write.cc:1004]: Can you add a comment to explain why?

File: db/db_impl/db_impl_write.cc [hunk 9]
```diff
@@ -1925,10 +1926,10 @@ Status DB::DeleteRange(const WriteOptions& opt,
                        const Slice& begin_key, const Slice& end_key) {
   WriteBatch batch;
   Status s = batch.DeleteRange(column_family, begin_key, end_key);
-  if (!s.ok()) {
-    return s;
+  if (s.ok()) {
+    s = Write(opt, &batch);
   }
-  return Write(opt, &batch);
+  return s;
```
Comment (siying) [inline db/db_impl/db_impl_write.cc:1932]: I don't see a difference between the changes.

**Review by siying — COMMENTED** [review]
My comments have been addressed. Thanks!

File: db/db_impl/db_impl_write.cc (hunk 8 above)
Comment (riversand963) [reply db/db_impl/db_impl_write.cc:1005]: Done.

File: db/db_impl/db_impl_write.cc (hunk 9 above)
Comment (riversand963) [reply db/db_impl/db_impl_write.cc]: Updated.
